test/object-store/sync/client_reset.cpp

```diff
@@ -6544,4 +6763,885 @@ TEST_CASE("client reset with nested collection", "[client reset][local][nested c
+                ndict->insert("key1", "value");
+            })
+            ->on_post_reset([&](SharedRealm local_realm) {
+                // Result: {"_id": <id>, "any_mixed": {{"key1": "value"}}}
+                advance_and_notify(*local_realm);
+                TableRef table = get_table(*local_realm, "TopLevel");
+                REQUIRE(table->size() == 1);
+                auto obj = table->get_object_with_primary_key(pk_val);
+                auto col = table->get_column_key("any_mixed");
+                // In the recovery case, the local instructions cannot be recovered
+                // because the path does not exist anymore.
+                object_store::Dictionary dict{local_realm, obj, col};
+                REQUIRE(dict.size() == 1);
+                REQUIRE(dict.get_any("key1").get_string() == "value");
+            })
+            ->run();
+    }
+    SECTION("Accessing ambiguous index triggers list copy") {
+        ObjectId pk_val = ObjectId::gen();
+        SyncTestFile config2(oas.app()->current_user(), "default");
+        config2.schema = config.schema;
+        auto test_reset = reset_utils::make_fake_local_client_reset(config, config2);
+        test_reset
+            ->setup([&](SharedRealm realm) {
+                // Baseline: {"_id": <id>, "any_mixed": {{"key1": [1, [2]]}, {"key2": 42}}}
+                auto table = get_table(*realm, "TopLevel");
+                auto obj = table->create_object_with_primary_key(pk_val);
+                auto col = table->get_column_key("any_mixed");
+                obj.set_collection(col, CollectionType::Dictionary);
+                object_store::Dictionary dictionary{realm, obj, col};
+                dictionary.insert_collection("key1", CollectionType::List);
+                dictionary.insert("key2", Mixed{42});
+                auto nlist = dictionary.get_list("key1");
+                nlist.add(Mixed{1});
+                nlist.insert_collection(1, CollectionType::List);
+                nlist = nlist.get_list(1);
+                nlist.add(Mixed{2});
+            })
+            ->make_local_changes([&](SharedRealm local_realm) {
+                // Local client: {"_id": <id>, "any_mixed": {{"key1": [1, [2, 3]]}, {"key2": 42}}}
+                advance_and_notify(*local_realm);
+                TableRef table = get_table(*local_realm, "TopLevel");
+                REQUIRE(table->size() == 1);
+                auto obj = table->get_object_with_primary_key(pk_val);
+                auto col = table->get_column_key("any_mixed");
+                // this insert operation triggers the list copy because the index becomes ambiguous
+                auto nlist = obj.get_list_ptr<Mixed>({col, "key1", 1});
+                nlist->add(Mixed{3});
+            })
+            ->make_remote_changes([&](SharedRealm remote_realm) {
+                // Remote client: {"_id": <id>, "any_mixed": {{"key1": [1, [2]]}, {"key2": 43}}}
```

Comment:
making a remote modification to the same nested list would make the end result more apparent that the remote changes are being overwritten

Reply:
good point. The reason I update `key2` is to actually show that only the list is overwritten (the previous implementation would overwrite everything top-down from the column key)